extensions/rocksdb-repos/RocksDatabase.cpp

```diff
@@ -0,0 +1,138 @@
+  rocksdb::Status result = impl_->Put(options, key, value);
+  if (result == rocksdb::Status::NoSpace()) {
+    db_->invalidate();
+  }
+  return result;
+}
+
+rocksdb::Status OpenRocksDB::Get(const rocksdb::ReadOptions& options, const rocksdb::Slice& key, std::string* value) {
+  rocksdb::Status result = impl_->Get(options, key, value);
+  if (result == rocksdb::Status::NoSpace()) {
+    db_->invalidate();
+  }
+  return result;
+}
+
+std::vector<rocksdb::Status> OpenRocksDB::MultiGet(const rocksdb::ReadOptions& options, const std::vector<rocksdb::Slice>& keys, std::vector<std::string>* values) {
+  std::vector<rocksdb::Status> results = impl_->MultiGet(options, keys, values);
+  for (const auto& result : results) {
+    if (result == rocksdb::Status::NoSpace()) {
+      db_->invalidate();
+    }
+  }
+  return results;
+}
+
+rocksdb::Status OpenRocksDB::Write(const rocksdb::WriteOptions& options, rocksdb::WriteBatch* updates) {
+  rocksdb::Status result = impl_->Write(options, updates);
+  if (result == rocksdb::Status::NoSpace()) {
+    db_->invalidate();
+  }
+  return result;
+}
+
+rocksdb::Status OpenRocksDB::Delete(const rocksdb::WriteOptions& options, const rocksdb::Slice& key) {
+  rocksdb::Status result = impl_->Delete(options, key);
+  if (result == rocksdb::Status::NoSpace()) {
+    db_->invalidate();
+  }
+  return result;
+}
+
+rocksdb::Status OpenRocksDB::Merge(const rocksdb::WriteOptions& options, const rocksdb::Slice& key, const rocksdb::Slice& value) {
+  rocksdb::Status result = impl_->Merge(options, key, value);
+  if (result == rocksdb::Status::NoSpace()) {
+    db_->invalidate();
+  }
+  return result;
+}
+
+bool OpenRocksDB::GetProperty(const rocksdb::Slice& property, std::string* value) {
+  return impl_->GetProperty(property, value);
+}
+
+std::unique_ptr<rocksdb::Iterator> OpenRocksDB::NewIterator(const rocksdb::ReadOptions& options) {
+  return std::unique_ptr<rocksdb::Iterator>{impl_->NewIterator(options)};
+}
+
+rocksdb::Status OpenRocksDB::NewCheckpoint(rocksdb::Checkpoint **checkpoint) {
+  return rocksdb::Checkpoint::Create(impl_.get(), checkpoint);
+}
+
+rocksdb::Status OpenRocksDB::FlushWAL(bool sync) {
+  rocksdb::Status result = impl_->FlushWAL(sync);
+  if (result == rocksdb::Status::NoSpace()) {
+    db_->invalidate();
+  }
+  return result;
+}
+
+rocksdb::DB* OpenRocksDB::get() {
+  return impl_.get();
+}
+
+RocksDatabase::RocksDatabase(const rocksdb::Options& options, const std::string& name, Mode mode) : open_options_(options), db_name_(name), mode_(mode) {}
+
+void RocksDatabase::invalidate() {
+  std::lock_guard<std::mutex> db_guard{ mtx_ };
+  // discard our own instance
+  impl_.reset();
+}
+
+utils::optional<OpenRocksDB> RocksDatabase::open() {
+  std::lock_guard<std::mutex> db_guard{ mtx_ };
+  if (!impl_) {
+    // database is not opened yet
+    rocksdb::DB* db_instance = nullptr;
+    rocksdb::Status result;
+    if (mode_ == Mode::ReadWrite) {
```

Comment:
I would use a `switch` so that we'll get a compiler warning if a new enum value is added and we forget to handle it.  But I realize that both of these events are unlikely, so I don't insist. :)

Reply:
could we promote such a warning to an error at the file/function/block level?

Reply:
🤔 I think we should enable that error, although not in this PR